### ISSUE TITLE
Particles speeds need to be normalized to the grid

### ISSUE THREAD

**Issue by mpiannucci** [asascience/ott-coops#281]
Particle movements are calculated within the grid the size of the bounds of the layer, but are not normalized. This results in global layers and local layers needing drastically looking defaults to represent winds or currents of the same speed. These need to be normalized to provide a consistent experience across layers.

**Comment by mpiannucci** [comment]
For Test: 

This looks bad by default for a wind layer: https://dev.oceansmap.com/maracoos-next/?permalink=N4IgtghgDgQjD2APEAuUB9Azgd1aANgHYDmqAtABwCsAdAEwCcVAzAIx0AsD33AbAOwAaEPggAXVMw41eFVrypUGs+VXkBfYekIBTPCJLle0hgAYOs-lx5Nmw0RJQc6MhlLq8GXOj4r91muDQAMpiAJ74emggAJYAJqhBUABGSAAyMcQAFhLChBBgUSAZ2WIABADiAE4QYSDCYlkArmDJ+TH4iTliUJgoAPT90DE0kClINADG8GD9mOGRmP0Abqz9Y6mI-fiZOWSrpv1iHTpLdFS8-YfX9SDzEUUbSANzC6fr0Jvbu2L7rKYgQLYMCYYI6MTHEiYfRVJrYKoAM2YAGswOhsDFCHFodF7pFElAIFVjpNFrdCcSYqSdABheD4eBVHEgADEzFM7GYMEELIsvBgAFE7GyAGJwAXc3kUAWeOg8gCCzBFABFmMqeQLlcrTMx5TyRbJTHRhcrjaZTFQefx5RadYDApMmlUqjpCGIALLQAAqMUK8x0UESdCNdDIHLIdAoXtYVBQ5rjpho5tMAC1bmMfX6xAGAEoQEhRADaIGDPjDrAjvC9FoTCaT5rTwlLofD7Orsfj5vrqZAAF1hI7na6xALIstxDF4IRUP9hE1CDExMEwv6wIlCmIqlTbphGdm4mlajoqgBJZU44uw+FI1HozHY27TCDzHa6aG99RAA

But GFS looks better but not perfect with the same defaults: 
http://20.85.242.58/?permalink=N4IgtghgDgQjD2APEAuUB9Azgd1aANgHYDmqAtABwCsAdAEwCcVAzAIx0AsD33AbAOwAaEPggAXVMw41eFVrypUGs+VXkBfYekIBTPCJLle0hgAYOs-lx5Nmw0RJQc6MhlLq8GXOj4r91muDQAMpiAJ74emggAJYAJqggEABOAMYA4jGYACrwUPDEydAAFjGpIMKEEGBRILn5hSVlFSBixQCuYABGVTH4icViYlCYKAD0Y5g6yQBu0zQpqcRZ8IT4Mbo0qfBgY4vLmGPJOphik9MzZSdjAOrwyfhx6PXw6ACy0GMfUMEX02NiPo6MamEEglqnCJRUBzZKYGKrVAUYSYeDtNInfRiPIFIpQUrlaLhKC1IqnaYtQGRTCoADaIEGw1GEyms3m+xWaw2Oi2Oz2aQORxOZ1Zl1S1zuDyeL3en2+vzZyQBQLGwAAXupVWFNcBEOoQABdYRUnTBGJqqJ0Ki8YTiMTJGJddqAxEoEAgAL2CBhaY0lC00DxRLYhp4gmUsIkxJksQUlFojHBnGNfHNYRgDZq+A7VCmdMQRBZnMoHzqA2ekDYMCYX5DDbEP2gYj4eBdCD4dDEABmmHQzCe2A2CWikMiiSgKUBqWpLQnySnkQAwvAW3DEgBiZimdjMGCCdcWXgwACidk3ADE4Me9weKMfPHR9wBBZjngAizDf++Pb7fpmYT77uesimHQZ5vmBpimFQ+78E+0H-h6gSpOixyEGI3zZDENTklAiR0KBdBkNuZB0BQ2SsFQKBQdRpg0FBpgAFotJAUBYThsZQAAShAJBRPSBE+MRrCkbw2TQbRtH0VBzHCIJREkVu4lUTRUHSUxhrCChyRoWIx6RDM4gIoQqCsHmIDtIQMRiMEYTkmAiQ1PaaYgKi846HEAAy3rTAAkm+fr0s2rbtp2PZ9gOQ4tNsECnOsug0uWQA

These are showing like the same data values so they should be able to represent the data equally

**Comment by mpiannucci** [comment]
Anf for reference, here is a good looking bit of settings for reference: http://20.85.242.58/?permalink=N4IgtghgDgQjD2APEAuUB9Azgd1aANgHYDmqAtAEwCsA7AHQDMAHACwAMTTAnAIzs9MKAGhD4IAF3I0edJgy4A2eQLY0qbLkwC+I9IQCmeUSVQ8aFOhQVKu1LlzbWaCkWMkoaDOjU9UmCrhoNJjYWBi0dcGgAZXEAT3xDNBAASwATVCioACMkaIl9fHwU8UMRQggwJJAASUhifQAnOJARcQALAFcwbIqU-Ez28XEoTBQAenHoFLpIHKQ6AGN4MHHMeMTMcYA3HnG53MQ1gqKS-TJtrnHxfv0t6gVxtien1pB1hOqDpAm1jbv9tBDsdSqdShcuCBItgwJhovphikSJgjMR8PBshB8OhiAAzTDoBhpdDYJEZZJQCCNG6LRIAEX0hEwJRaKB4thElOpKVp+miUH0+nJbEYnKpNMSACUIIQ0isDJgUSgRWw2MJ3v9MlyJXc3tqeYkAMLwdGNJUgADEDDYPAoDBgQgtLGsMAAogxHQwAGJwV0Op1MV0BYQWgCC3rpDDpjtddLpbAYocdXoUITtjrpdtVVEdNFDbHUDD14oN+gAMoziB1UBQeGLubzoikAF5JGRUCIiRadRqNRniACy0AAKikqut9FBMhQ1RQyDbKExhzwqMq2Gu6Kq2AAtN5zUfj0pQaUkJIAbRAM4oc4XVmHBbXG63u5EV5vPDI1vvq9VT9Vu4AXS7Hs+0IcRXUSbYJBSeBCFMNgRE6QgSmiOIJzATIqnERoeTeTB4GpIUywgOImhqOklQvNEMSxHF8UJYlSVlN5lggdZigVEAAK0IA

**Comment by ndellicarpini** [comment]
So to start I'll try and make the density a function of the bbox area, and then from there we can figure out if other aspects should be also dependent on map/layer attributes.

**Comment by ndellicarpini** [comment]
actually scratch that, I think I'm going to make the density contingent on the current map zoom. I think from a user perspective it makes more sense that the density they set is universal regardless of zoom level, but we'll see if I run into re-rendering problems.

Edit: this is probably going to take some serious optimizations if I want to do this, I'll mess with that too.

**Comment by mpiannucci** [comment]
https://github.com/mapbox/webgl-wind

https://mapbox.github.io/webgl-wind/demo/

https://blog.mapbox.com/how-i-built-a-wind-map-with-webgl-b63022b5537f

**Comment by ndellicarpini** [comment]
After looking at those links, I have a much better understanding of how you are using the framebuffer and why, and it makes a lot of sense in its current implementation. If you can, I'd like to chat soon about trying to make particle density somehow contingent on map zoom level, but that's realistically outside the scope of this issue.

As for this issue, I'm thinking of just making particle speed a function of the ratio between layer geo_extents and global latlng minmax. I'll play with how strong to make the size scaling of the particle speed, and I can post some demo videos here if that approach makes sense.

**Comment by jamesdoyle21** [comment]
Sounds good, Nick. And yes, I agree that particle density should be it's own issue. Not a high priority right now, but when you are ready to look at it and write out you ideas, go ahead and make a separate issue for it

**Comment by mpiannucci** [comment]
Awesome nick, that sounds like a great approach. After you get the speed right we can chat about the density

**Comment by ndellicarpini** [comment]
OK so here is a video. I was basically able to eyeball an algorithm for matching the speeds at different layer sizes. I think the different particles go the same speed, but I would like any opinions if you believe the speeds don't match. I also eyeballed adjusting the randomness, but there was no way to really calculate an algorithm for that since I can't match the randomness visually. 

At the start of the video, the layers are activated with the new default settings I think we should use. Let me know your thoughts on any aspect of the video.

[I think only rps people can click this?](https://rpsgroup-my.sharepoint.com/:v:/p/nick_dellicarpini/EbMcp1GmckBOiOPRO0k-zfgBfbH1C8N-oZOq7vaYGams8Q?e=1eQNX8)

**Comment by ndellicarpini** [comment]
alright video's up

**Comment by jamesdoyle21** [comment]
This looks good. But testing it really showed that nailing default settings for particles for different layers is going to be hard. Something to keep in mind for future work.
